python/ray/data/_internal/execution/interfaces/op_runtime_metrics.py

```diff
@@ -0,0 +1,189 @@
+from dataclasses import dataclass, field, fields
+from typing import Dict, Optional
+
+import ray
+from ray.data._internal.execution.interfaces.ref_bundle import RefBundle
+from ray.data._internal.memory_tracing import trace_allocation
+
+
+@dataclass
+class RunningTaskInfo:
+    inputs: RefBundle
+    num_outputs: int
+    bytes_outputs: int
+
+
+@dataclass
+class OpRuntimeMetrics:
+    """Runtime metrics for a PhysicalOperator."""
+
+    # === Inputs-related metrics ===
+
+    # Number of received input blocks.
+    num_inputs_received: int = 0
+    # Total size in bytes of received input blocks.
+    bytes_inputs_received: int = 0
+
+    # Number of processed input blocks.
+    num_inputs_processed: int = 0
+    # Total size in bytes of processed input blocks.
+    bytes_inputs_processed: int = 0
+
+    # === Outputs-related metrics ===
+
+    # Number of generated output blocks.
+    num_outputs_generated: int = 0
+    # Total size in bytes of generated output blocks.
+    bytes_outputs_generated: int = 0
```

Comment:
We're also planning on reporting the number of rows outputted, would it be possible to add it to this class?

Reply:
num_rows in Block is now nullable. I'd like to make it non-none after migrating to this new API. Need to confirm how big the change will be. if it turns out to need a big change, I'll probably do that in a different PR.